191 remove treenode and use treegoal for later development

aoxiangx opened the pull request:
every commits have detail stated.
all tests passed, 58 in total (deprecated 1 and also added a new 1)

aoxiangx committed:
refactor(types): introduce TreeGoal as unified type, deprecate TreeNode

- Add TreeGoal as the primary tree structure type (using `id` field)
- Mark TreeNode as deprecated alias for TreeGoal
- Mark newTreeItem as deprecated alias for newTreeGoal
- Move InstanceId and Label to top of file to avoid circular refs
- Replace TreeNode["instanceId"] with InstanceId throughout

This is Step 1 of removing normalization - TreeNode will be fully
removed in subsequent steps.

aoxiangx committed:
refactor(treeDataSlice): use TreeGoal directly in Redux state

- Change state.tree type from TreeNode[] to TreeGoal[]
- Replace goalId with id throughout (TreeGoal uses id)
- Simplify createTreeFromTreeData to identity function (deprecated)
- Remove createTreeDataFromTreeNode usage in selector
- Add deprecated aliases: newTreeNode, findTreeNodeByInstanceId

This is Step 2 of removing normalization - Redux now stores
TreeGoal directly without conversion.

aoxiangx committed:
refactor(FileProvider): remove conversion functions, use TreeGoal directly

- Replace TreeItem/TreeNode with TreeGoal throughout FileProvider
- Simplify createTreeDataFromTreeNode to identity (deprecated)
- Simplify convertTreeDataToClusters - no longer needs goals param
- Store TreeGoal[] directly to localStorage without conversion
- Remove unnecessary data transformations

aoxiangx committed:
refactor: replace TreeNode with TreeGoal, remove deprecated aliases

Update all references:
- GraphUtils.tsx: TreeItem/TreeNode → TreeGoal/InstanceId
- GraphHelpers.tsx: TreeNode → InstanceId

 Cleanup:
- Remove TreeNode and newTreeItem aliases from types.ts
- Remove newTreeNode and createTreeFromTreeData from treeDataSlice.ts
- Remove createTreeDataFromTreeNode from FileProvider.tsx
- Remove findTreeNodeByInstanceId alias

aoxiangx committed:
refactor: replace TreeItem with TreeGoal in components and data

- Update initialTabs.ts: TreeItem → TreeGoal, newTreeItem → newTreeGoal
- Update GoalList, GoalListTable, SectionPanel, Tree, WelcomeButtons
- Update SidebarItems: newTreeItem → newTreeGoal

aoxiangx committed:
test: update tests to use TreeGoal instead of TreeNode/TreeItem

- Replace newTreeItem/TreeItem/TreeNode with newTreeGoal/TreeGoal/InstanceId
- Replace newTreeNode with createTreeGoalNode
- Replace findTreeNodeByInstanceId with findTreeGoalByInstanceId
- Remove createTreeFromTreeData test (function removed)
- Update assertions: goalId → id

All 57 tests pass.

aoxiangx committed:
fix: update tree node content when editing goal via canvas double-click

- updateTextForInstanceId now also updates tree node content
- Previously only state.goals was updated, but tree stores full TreeGoal
- Add test for canvas double-click edit functionality

Fixes goal name editing on canvas not persisting to tree hierarchy.

aoxiangx committed:
Merge remote-tracking branch 'origin/develop' into 191-remove-treenode-and-use-treegoal-for-later-development

aoxiangx commented:
Updated and merged with the changes from latest develop branch, waiting for review.

bggolding commented on src/components/types.ts:
```diff
@@ -1,10 +1,16 @@
 export interface GoalBase {
     GoalID: number
-    instanceId: TreeNode["instanceId"]
+    instanceId: InstanceId
```
I would like to see this kept as `TreeGoal["instanceId"]`. That way,  if the definition of that type in`TreeGoal` changes, all the references to it will pick up that change automatically.

bggolding commented on src/components/types.ts:
```diff
@@ -117,41 +123,39 @@ export const GoalModelProjectSchema = z.object({
+// ============================================
+// TreeGoal - the unified tree structure type
+// ============================================
+export type TreeGoal = {
```
Does this comment really add anything?

bggolding commented on src/components/types.ts:
```diff
@@ -117,41 +123,39 @@ export const GoalModelProjectSchema = z.object({
+// ============================================
+/** @deprecated Use TreeGoal instead */
+export type TreeNode = TreeGoal;
+/** @deprecated Use newTreeGoal instead */
```
Just delete it. Don't mark it as deprecated, just get it out of the code. The place for history in the project is in the change sets in the git repo, not in the code.

bggolding commented on src/components/context/treeDataSlice.ts:
```diff
@@ -1,52 +1,55 @@
-        instanceId?: TreeItem["instanceId"];
-        children?: TreeNode[];
+        id: TreeGoal["id"];
+        instanceId?: InstanceId;
```
Use `TreeGoal["instanceId"]`.

bggolding commented on src/components/context/treeDataSlice.ts:
```diff
@@ -1,52 +1,55 @@
-};
+
+/** @deprecated No longer needed - tree now stores TreeGoal directly */
+export const createTreeFromTreeData = (treeData: TreeGoal[]): TreeGoal[] => treeData;
```
If it's no longer needed, delete it. Don't leave old cruft in the code -- it just makes more stuff people have to learn about to work in the codebase.

bggolding commented on src/components/context/treeDataSlice.ts:
```diff
@@ -130,80 +133,87 @@ const generateMaxSuffix = (treeIds: Record<TreeItem["id"], TreeItem["instanceId"
         tabs,
         goals,
-        tree: createTreeFromTreeData(treeData),
+        tree: treeData,  // No conversion needed - tree now stores TreeGoal directly
```
The comment isn't necessary or helpful.

bggolding commented on src/components/context/treeDataSlice.ts:
```diff
@@ -230,7 +240,7 @@ export const treeDataSlice = createSlice({
         },
         updateTextForInstanceId: (state, action: PayloadAction<{
-            instanceId: string,
+            instanceId: string,  // Accept string for compatibility with getCellNumericIds
```
I would not do this. Either make the user cast the type to do this assignment or (better) write a little function that does a test that the InstanceId conforms to the correct pattern.

Something like:
```typescript
const checkInstanceId = (id: string): InstanceId => {
    // raise exception if id is in wrong format
    return id as InstanceId;
}
```

bggolding commented on src/components/context/FileProvider.tsx:
```diff
@@ -192,28 +173,25 @@ const FileProvider: React.FC<PropsWithChildren> = ({children}) => {
+            treeData: state.tree,  // No conversion needed
             tabData: computedTabData,
-            cluster: convertTreeDataToClusters(state.goals, state.tree),
+            cluster: convertTreeDataToClusters(state.tree),  // Simplified - no goals param needed
```
Not a useful comment. You are documenting how the code used to be, not how it is now.

bggolding commented on src/components/context/FileProvider.test.tsx:
```diff
@@ -118,51 +118,51 @@ describe('FileProvider', () => {
         const treeData = [
-                newTreeItem({type: "Do", id: 1, children: [
-                    newTreeItem({type: "Do", id: 2})
+                newTreeGoal({type: "Do", id: 1, children: [
```
The indenting here looks not quite right.

bggolding commented on src/components/context/treeDataSlice.test.ts:
```diff
@@ -155,82 +154,71 @@ describe('treeDataSlice', () => {
+        const nonExistentId: InstanceId = "999-999";
+        expect(findTreeGoalByInstanceId(testTree, nonExistentId)).toBeUndefined();
     });
-});
```
Thank you for updating the unit tests.

bggolding reviewed (changes requested):
Looks pretty good but I'd like to see some changes.

aoxiangx committed:
refactor: remove unnecessary explanatory comments

Remove comments that document historical changes or state obvious facts.

aoxiangx committed:
refactor: use TreeGoal["instanceId"] for type consistency

Replace direct InstanceId references with TreeGoal["instanceId"] to ensure
type changes propagate automatically. Using indexed access types ensures
that if TreeGoal's instanceId type changes, all references update automatically.

aoxiangx committed:
feat: add validateInstanceId helper for type safety

Add validation function instead of casting strings to InstanceId,
providing runtime type checking. Runtime validation is safer than
casting and throws clear errors for invalid input.

aoxiangx commented:
The changes have been made based on your advice.
Looking forward to your review.

bggolding commented on src/components/utils/GraphUtils.tsx:
```diff
@@ -185,12 +185,24 @@ export function generateCellId<T extends keyof IdsForType>(type: T, ids: IdsForT
 }
 
+export const validateInstanceId = (id: string): InstanceId => {
+    const bits = id.split("-").map(s => s.trim());
```
This is all looking really good but if the instanceId has spaces in it then it's badly formatted and the code should complain.

Would it be better to use a regular expression to do this checking rather than pulling the instanceId to pieces? I seem to recall that there might already be a reg exp to do that somewhere.

bggolding replied:
This should be a separate issue, not done as part of this one.

bggolding reviewed (changes requested):
This is looking really quite good now. Thanks for all your work.

I am starting to wonder if all the pulling apart and validation of instanceId should be done by a reg exp rather than in code. It makes it easier to understand and validate.
